File: csharp/Assembler/TemplatePerformance/PerformanceUtils.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using Assembler.TemplateEngine;
using Assembler.TemplateLoader;

namespace Assembler.TemplatePerformance;

public static class PerformanceUtils
{
    public class PerfSummaryRow
    {
        public string? AppSite { get; set; }
        public string? AppFile { get; set; }
        public string? AppView { get; set; }
        public int Iterations { get; set; }
        public long NormalTimeTicks { get; set; }
        public long PreProcessTimeTicks { get; set; }
        public int OutputSize { get; set; }
        public string? ResultsMatch { get; set; }
        public string? PerfDifference { get; set; }
        
        // Helper properties for display
        public double NormalTimeMs => (double)NormalTimeTicks / Stopwatch.Frequency * 1000;
        public double PreProcessTimeMs => (double)PreProcessTimeTicks / Stopwatch.Frequency * 1000;
    }

    public static List<PerfSummaryRow> RunPerformanceComparison(string assemblerWebDirPath, bool enableJsonProcessing, string? appSiteFilter = null, bool skipDetails = false)
    {
        int iterations = 1000;
        string appSitesPath = Path.Combine(assemblerWebDirPath, "AppSites");
        var summaryRows = new List<PerfSummaryRow>();
        if (!Directory.Exists(appSitesPath))
            return summaryRows;
        var allAppSiteDirs = Directory.GetDirectories(appSitesPath, "*", SearchOption.TopDirectoryOnly);
        var appSites = allAppSiteDirs.Select(dir => Path.GetFileName(dir)).ToArray();
        foreach (var testAppSite in appSites)
        {
            if (!string.IsNullOrEmpty(appSiteFilter) && !testAppSite.Equals(appSiteFilter, StringComparison.OrdinalIgnoreCase))
                continue;
            string appSiteDir = Path.Combine(appSitesPath, testAppSite);
            var htmlFiles = Directory.GetFiles(appSiteDir, "*.html", SearchOption.TopDirectoryOnly);
            foreach (var htmlFilePath in htmlFiles)
            {
                var appFileName = Path.GetFileNameWithoutExtension(htmlFilePath);
                try
                {
                    LoaderNormal.ClearCache();
                    LoaderPreProcess.ClearCache();
                    var templates = LoaderNormal.LoadGetTemplateFiles(assemblerWebDirPath, testAppSite);
                    var siteTemplates = LoaderPreProcess.LoadProcessGetTemplateFiles(assemblerWebDirPath, testAppSite);
                    if (templates == null || templates.Count == 0)
                    {
                        if (!skipDetails) Console.WriteLine($"❌ No templates found for {testAppSite}");
                        continue;
                    }
                    var mainTemplateKey = (testAppSite + "_" + appFileName).ToLowerInvariant();
                    if (!templates.TryGetValue(mainTemplateKey, out var mainTemplate))
                    {
                        if (!skipDetails) Console.WriteLine($"❌ No main template found for {mainTemplateKey}");
                        continue;
                    }
                    if (!skipDetails)
                    {
                        Console.WriteLine($"Template Key: {mainTemplateKey}");
                        Console.WriteLine($"Templates available: {templates.Count}");
                    }
                    var appViewScenarios = new List<(string AppView, string AppViewPrefix)> { ("", "") };
                    var viewsPath = Path.Combine(appSiteDir, "Views");
                    if (Directory.Exists(viewsPath))
                    {
                        var viewFiles = Directory.GetFiles(viewsPath, "*.html");
                        foreach (var viewFile in viewFiles)
                        {
                            var viewName = Path.GetFileNameWithoutExtension(viewFile);
                            var appView = "";
                            var appViewPrefix = "";
                            if (viewName.ToLowerInvariant().Contains("content"))
                            {
                                var contentIndex = viewName.ToLowerInvariant().IndexOf("content");
                                if (contentIndex > 0)
                                {
                                    var viewPart = viewName.Substring(0, contentIndex);
                                    if (viewPart.Length > 0)
                                    {
                                        appView = char.ToUpper(viewPart[0]) + viewPart.Substring(1);
                                        appViewPrefix = appView.Substring(0, Math.Min(appView.Length, 6));
                                    }
                                }
                            }
                            if (!string.IsNullOrEmpty(appView))
                                appViewScenarios.Add((appView, appViewPrefix));
                        }
                    }
                    foreach (var scenario in appViewScenarios)
                    {
                        if (!skipDetails)
                        {
                            Console.WriteLine(new string('-', 60));
                            Console.WriteLine($">>> C# SCENARIO : '{testAppSite}', '{appFileName}', '{scenario.AppView}', '{scenario.AppViewPrefix}'");
                            Console.WriteLine($"Iterations per test: {iterations:N0}");
                        }
                        var normalEngine = new EngineNormal();
                        normalEngine.AppViewPrefix = scenario.AppViewPrefix;
                        
                        // JIT Warmup - run a few iterations first to warm up the JIT
                        for (int warmup = 0; warmup < 100; warmup++)
                        {
                            normalEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, templates, enableJsonProcessing);
                        }
                        
                        var sw = Stopwatch.StartNew();
                        string resultNormal = "";
                        for (int i = 0; i < iterations; i++)
                        {
                            resultNormal = normalEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, templates, enableJsonProcessing);
                        }
                        sw.Stop();
                        var normalTime = sw.ElapsedMilliseconds;
                        var normalTicks = sw.ElapsedTicks;
                        if (!skipDetails)
                        {
                            Console.WriteLine($"[Normal Engine] {iterations:N0} iterations: {normalTime}ms ({normalTicks:N0} ticks)");
                            Console.WriteLine($"[Normal Engine] Avg: {(double)normalTime / iterations:F3}ms per op, Output size: {resultNormal.Length} chars");
                        }
                        LoaderNormal.ClearCache();
                        LoaderPreProcess.ClearCache();
                        var preProcessEngine = new EnginePreProcess();
                        preProcessEngine.AppViewPrefix = scenario.AppViewPrefix;
                        
                        // JIT Warmup for PreProcess engine
                        for (int warmup = 0; warmup < 100; warmup++)
                        {
                            preProcessEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, siteTemplates.Templates, enableJsonProcessing);
                        }
                        
                        sw.Restart();
                        string resultPreProcess = "";
                        for (int i = 0; i < iterations; i++)
                        {
                            resultPreProcess = preProcessEngine.MergeTemplates(testAppSite, appFileName, scenario.AppView, siteTemplates.Templates, enableJsonProcessing);
                        }
                        sw.Stop();
                        var preProcessTime = sw.ElapsedMilliseconds;
                        var preProcessTicks = sw.ElapsedTicks;
                        if (!skipDetails)
                        {
                            Console.WriteLine($"[PreProcess Engine] {iterations:N0} iterations: {preProcessTime}ms ({preProcessTicks:N0} ticks)");
                            Console.WriteLine($"[PreProcess Engine] Avg: {(double)preProcessTime / iterations:F3}ms per op, Output size: {resultPreProcess.Length} chars");
                            Console.WriteLine($">>> C# PERFORMANCE COMPARISON:");
                            Console.WriteLine(new string('-', 50));
                            var difference = preProcessTime - normalTime;
                            var differencePercent = normalTime > 0 ? ((double)difference / normalTime) * 100 : 0;
                            var tickDifference = preProcessTicks - normalTicks;
                            var tickDifferencePercent = normalTicks > 0 ? ((double)tickDifference / normalTicks) * 100 : 0;
                            Console.WriteLine($"Time difference: {difference}ms ({differencePercent:F1}%)");
                            Console.WriteLine($"Tick difference: {tickDifference:N0} ticks ({tickDifferencePercent:F1}%)");
                            Console.WriteLine($"Results match: {(resultNormal == resultPreProcess ? "✅ YES" : "❌ NO")}");
                        }
                        summaryRows.Add(new PerfSummaryRow
                        {
                            AppSite = testAppSite,
                            AppFile = appFileName,
                            AppView = scenario.AppView,
                            Iterations = iterations,
                            NormalTimeTicks = normalTicks,
                            PreProcessTimeTicks = preProcessTicks,
                            OutputSize = resultNormal.Length,
                            ResultsMatch = (resultNormal == resultPreProcess ? "YES" : "NO"),
                            PerfDifference = normalTicks > 0 ? $"{((double)(preProcessTicks - normalTicks) / normalTicks * 100):F1}%" : "0%"
                        });
                    }
                }
                catch (Exception ex)
                {
                    if (!skipDetails) Console.WriteLine($"❌ Error in performance testing for {testAppSite}: {ex.Message}");
                }
            }
        }
        return summaryRows;
    }

    public static void PrintPerfSummaryTable(string assemblerWebDirPath, List<PerfSummaryRow> summaryRows)
    {
        if (summaryRows == null || summaryRows.Count == 0)
            return;
        Console.WriteLine("\n==================== C# PERFORMANCE SUMMARY ====================\n");

        var headers = new[] { "AppSite", "AppView", "Normal(ms)", "PreProc(ms)", "Match", "PerfDiff" };
        int colCount = headers.Length;
        int[] widths = new int[colCount];
        for (int i = 0; i < colCount; i++)
        {
            widths[i] = headers[i].Length;
        }
        foreach (var row in summaryRows)
        {
            widths[0] = Math.Max(widths[0], row.AppSite?.Length ?? 0);
            widths[1] = Math.Max(widths[1], row.AppView?.Length ?? 0);
            widths[2] = Math.Max(widths[2], row.NormalTimeMs.ToString("F2").Length);
            widths[3] = Math.Max(widths[3], row.PreProcessTimeMs.ToString("F2").Length);
            widths[4] = Math.Max(widths[4], row.ResultsMatch?.Length ?? 0);
            widths[5] = Math.Max(widths[5], row.PerfDifference?.Length ?? 0);
        }
        // Print header
        Console.Write("| ");
        for (int i = 0; i < colCount; i++)
        {
            Console.Write(headers[i].PadRight(widths[i]));
            if (i < colCount - 1) Console.Write(" | ");
        }
        Console.WriteLine(" |");
        // Print divider
        Console.Write("|");
        for (int i = 0; i < colCount; i++)
        {
            Console.Write(" " + new string('-', widths[i]) + " ");
            if (i < colCount - 1) Console.Write("|");
        }
        Console.WriteLine("|");
        // Print rows
        foreach (var row in summaryRows)
        {
            Console.Write("| ");
            Console.Write((row.AppSite ?? "").PadRight(widths[0]));
            Console.Write(" | ");
            Console.Write((row.AppView ?? "").PadRight(widths[1]));
            Console.Write(" | ");
            Console.Write(row.NormalTimeMs.ToString("F2").PadRight(widths[2]));
            Console.Write(" | ");
            Console.Write(row.PreProcessTimeMs.ToString("F2").PadRight(widths[3]));
            Console.Write(" | ");
            Console.Write((row.ResultsMatch ?? "").PadRight(widths[4]));
            Console.Write(" | ");
            Console.Write((row.PerfDifference ?? "").PadRight(widths[5]));
            Console.WriteLine(" |");
        }
        // Print bottom divider
        Console.Write("|");
        for (int i = 0; i < colCount; i++)
        {
            Console.Write(" " + new string('-', widths[i]) + " ");
            if (i < colCount - 1) Console.Write("|");
        }
        Console.WriteLine("|");

        // Save HTML file
        try
        {
            var html = new StringBuilder();
            html.AppendLine("<html><head><title>C# Performance Summary Table</title><style>table{border-collapse:collapse;}th,td{border:1px solid #888;padding:4px;}th{background:#eee;}</style></head><body>");
            html.AppendLine("<h2>C# Performance Summary Table</h2>");
            html.AppendLine("<table>");
            html.Append("<tr>");
            foreach (var h in headers) html.Append($"<th>{h}</th>");
            html.AppendLine("</tr>");
            foreach (var row in summaryRows)
            {
                html.Append("<tr>");
                html.Append($"<td>{row.AppSite}</td>");
                html.Append($"<td>{row.AppView}</td>");
                html.Append($"<td>{row.NormalTimeMs:F2}</td>");
                html.Append($"<td>{row.PreProcessTimeMs:F2}</td>");
                html.Append($"<td>{row.ResultsMatch}</td>");
                html.Append($"<td>{row.PerfDifference}</td>");
                html.AppendLine("</tr>");
            }
            html.AppendLine("</table></body></html>");

            var outFile = Path.Combine(assemblerWebDirPath, "csharp_perfsummary.html");
            File.WriteAllText(outFile, html.ToString());
            Console.WriteLine($"Performance summary HTML saved to: {outFile}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error saving performance summary HTML: {ex.Message}");
        }

        // Save JSON file
        try
        {
            var jsonFile = Path.Combine(assemblerWebDirPath, "csharp_perfsummary.json");
            var json = Arshu.App.Json.JsonConverter.SerializeObject(summaryRows, true);
            File.WriteAllText(jsonFile, json);
            Console.WriteLine($"Performance summary JSON saved to: {jsonFile}");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error saving performance summary JSON: {ex.Message}");
        }
    }
}

File: php/Assembler/src/TemplatePerformance/PerformanceUtils.php
<?php

namespace Assembler\TemplatePerformance;

use Assembler\TemplateLoader\LoaderNormal;
use Assembler\TemplateLoader\LoaderPreProcess;
use Assembler\TemplateEngine\EngineNormal;
use Assembler\TemplateEngine\EnginePreProcess;

class PerfSummaryRow
{
    public ?string $AppSite;
    public ?string $AppFile;
    public ?string $AppView;
    public int $Iterations;
    public float $NormalTimeMs;
    public float $PreProcessTimeMs;
    public int $OutputSize;
    public ?string $ResultsMatch;
    public ?string $PerfDifference;

    public function __construct(?string $appSite = null, ?string $appFile = null, ?string $appView = null, int $iterations = 0, float $normalTimeMs = 0.0, float $preProcessTimeMs = 0.0, int $outputSize = 0, ?string $resultsMatch = null, ?string $perfDifference = null)
    {
        $this->AppSite = $appSite;
        $this->AppFile = $appFile;
        $this->AppView = $appView;
        $this->Iterations = $iterations;
        $this->NormalTimeMs = $normalTimeMs;
        $this->PreProcessTimeMs = $preProcessTimeMs;
        $this->OutputSize = $outputSize;
        $this->ResultsMatch = $resultsMatch;
        $this->PerfDifference = $perfDifference;
    }
}

class PerformanceUtils
{
    /**
     * @param string $assemblerWebDirPath
     * @param bool $enableJsonProcessing
     * @param string|null $appSiteFilter
     * @param bool $skipDetails
     * @return PerfSummaryRow[]
     */
    public static function runPerformanceComparison(string $assemblerWebDirPath, bool $enableJsonProcessing, ?string $appSiteFilter = null, bool $skipDetails = false): array
    {
        $startTime = microtime(true);
        echo "\n========== RunPerformanceComparison ==========\n";
        $iterations = 1000;
        $appSitesPath = $assemblerWebDirPath . DIRECTORY_SEPARATOR . 'AppSites';
        $summaryRows = [];
        if (!is_dir($appSitesPath)) {
            return $summaryRows;
        }
        $allAppSiteDirs = array_filter(scandir($appSitesPath), function($dir) use ($appSitesPath) {
            return $dir !== '.' && $dir !== '..' && is_dir($appSitesPath . DIRECTORY_SEPARATOR . $dir);
        });
        $appSites = array_map('basename', $allAppSiteDirs);
        foreach ($appSites as $testAppSite) {
            if ($appSiteFilter !== null && strcasecmp($testAppSite, $appSiteFilter) !== 0) {
                continue;
            }
            $appSiteDir = $appSitesPath . DIRECTORY_SEPARATOR . $testAppSite;
            $htmlFiles = glob($appSiteDir . DIRECTORY_SEPARATOR . '*.html');
            foreach ($htmlFiles as $htmlFilePath) {
                $appFileName = basename($htmlFilePath, '.html');
                try {
                    LoaderNormal::clearCache();
                    LoaderPreProcess::clearCache();
                    $templates = LoaderNormal::loadGetTemplateFiles($assemblerWebDirPath, $testAppSite);
                    $siteTemplates = LoaderPreProcess::loadProcessGetTemplateFiles($assemblerWebDirPath, $testAppSite);
                    if ($templates === null || count($templates) === 0) {
                        continue;
                    }
                    $mainTemplateKey = strtolower($testAppSite . '_' . $appFileName);
                    if (!isset($templates[$mainTemplateKey])) {
                        continue;
                    }
                    $appViewScenarios = [['AppView' => '', 'AppViewPrefix' => '']];
                    $viewsPath = $appSiteDir . DIRECTORY_SEPARATOR . 'Views';
                    if (is_dir($viewsPath)) {
                        $viewFiles = glob($viewsPath . DIRECTORY_SEPARATOR . '*.html');
                        foreach ($viewFiles as $viewFile) {
                            $viewName = basename($viewFile, '.html');
                            $appView = '';
                            $appViewPrefix = '';
                            $contentIndex = stripos($viewName, 'content');
                            if ($contentIndex !== false && $contentIndex > 0) {
                                $viewPart = substr($viewName, 0, $contentIndex);
                                if (strlen($viewPart) > 0) {
                                    $appView = ucfirst($viewPart);
                                    $appViewPrefix = substr($appView, 0, min(strlen($appView), 6));
                                }
                            }
                            if (!empty($appView)) {
                                $appViewScenarios[] = ['AppView' => $appView, 'AppViewPrefix' => $appViewPrefix];
                            }
                        }
                    }
                    foreach ($appViewScenarios as $scenario) {
                        if (!$skipDetails) {
                            echo str_repeat('-', 60) . "\n";
                            echo "[PHP] Testing: AppSite={$testAppSite}, AppFile={$appFileName}, AppView={$scenario['AppView']}\n";
                            echo "[PHP] Iterations: " . number_format($iterations) . "\n";
                        }
                        LoaderNormal::clearCache();
                        LoaderPreProcess::clearCache();
                        $start = microtime(true);
                        $resultNormal = '';
                        $normalEngine = new EngineNormal();
                        $normalEngine->setAppViewPrefix($scenario['AppViewPrefix']);
                        for ($i = 0; $i < $iterations; $i++) {
                            $resultNormal = $normalEngine->mergeTemplates($testAppSite, $appFileName, $scenario['AppView'], $templates, $enableJsonProcessing);
                        }
                        $normalTime = round((microtime(true) - $start) * 1000, 2);
                        if (!$skipDetails) {
                            $avg = $normalTime / $iterations;
                            echo "[PHP] Normal Engine:     {$normalTime}ms | Avg: " . number_format($avg, 3) . "ms/op | Size: " . strlen($resultNormal) . " chars\n";
                        }
                        LoaderNormal::clearCache();
                        LoaderPreProcess::clearCache();
                        $preProcessEngine = new EnginePreProcess();
                        $preProcessEngine->setAppViewPrefix($scenario['AppViewPrefix']);
                        $start = microtime(true);
                        $resultPreProcess = '';
                        for ($i = 0; $i < $iterations; $i++) {
                            $resultPreProcess = $preProcessEngine->mergeTemplates($testAppSite, $appFileName, $scenario['AppView'], $siteTemplates->templates, $enableJsonProcessing);
                        }
                        $preProcessTime = round((microtime(true) - $start) * 1000, 2);
                        if (!$skipDetails) {
                            $avg = $preProcessTime / $iterations;
                            echo "[PHP] PreProcess Engine: {$preProcessTime}ms | Avg: " . number_format($avg, 3) . "ms/op | Size: " . strlen($resultPreProcess) . " chars\n";

                            $difference = round($preProcessTime - $normalTime, 2);
                            $differencePercent = $normalTime > 0 ? ($difference / $normalTime) * 100 : 0;
                            $signMs = $difference >= 0 ? '+' : '';
                            $signPct = $differencePercent >= 0 ? '+' : '';
                            $matchStr = $resultNormal === $resultPreProcess ? 'YES' : 'NO';
                            echo "[PHP] Performance: {$signMs}{$difference}ms ({$signPct}" . number_format($differencePercent, 1) . "%) | Match: {$matchStr}\n";
                        }
                        $summaryRows[] = new PerfSummaryRow(
                            $testAppSite,
                            $appFileName,
                            $scenario['AppView'],
                            $iterations,
                            $normalTime,
                            $preProcessTime,
                            strlen($resultNormal),
                            ($resultNormal === $resultPreProcess ? "YES" : "NO"),
                            $normalTime > 0 ? number_format(($preProcessTime - $normalTime) / $normalTime * 100, 1) . "%" : "0%"
                        );
                    }
                } catch (\Exception $ex) {
                    if (!$skipDetails) echo "❌ Error in performance testing for {$testAppSite}: {$ex->getMessage()}\n";
                }
            }
        }
        $elapsed = (microtime(true) - $startTime) * 1000;
        echo sprintf("\n========== Performance Testing Completed in %.0fms ==========\n\n", $elapsed);
        return $summaryRows;
    }

    /**
     * @param string $assemblerWebDirPath
     * @param PerfSummaryRow[] $summaryRows
     */
    public static function printPerfSummaryTable(string $assemblerWebDirPath, array $summaryRows): void
    {
        if (empty($summaryRows)) {
            return;
        }
        echo "\n==================== PHP PERFORMANCE SUMMARY ====================\n\n";

        $headers = ['AppSite', 'AppView', 'Normal(ms)', 'PreProc(ms)', 'Match', 'PerfDiff'];
        $colCount = count($headers);
        $widths = array_fill(0, $colCount, 0);
        for ($i = 0; $i < $colCount; $i++) {
            $widths[$i] = strlen($headers[$i]);
        }
        foreach ($summaryRows as $row) {
            $widths[0] = max($widths[0], strlen($row->AppSite ?? ''));
            $widths[1] = max($widths[1], strlen($row->AppView ?? ''));
            $widths[2] = max($widths[2], strlen(number_format($row->NormalTimeMs, 2)));
            $widths[3] = max($widths[3], strlen(number_format($row->PreProcessTimeMs, 2)));
            $widths[4] = max($widths[4], strlen($row->ResultsMatch ?? ''));
            $widths[5] = max($widths[5], strlen($row->PerfDifference ?? ''));
        }
        // Print header
        echo '| ';
        for ($i = 0; $i < $colCount; $i++) {
            echo str_pad($headers[$i], $widths[$i]);
            if ($i < $colCount - 1) echo ' | ';
        }
        echo " |\n";
        // Print divider
        echo '|';
        for ($i = 0; $i < $colCount; $i++) {
            echo ' ' . str_repeat('-', $widths[$i]) . ' ';
            if ($i < $colCount - 1) echo '|';
        }
        echo "|\n";
        // Print rows
        foreach ($summaryRows as $row) {
            echo '| ';
            echo str_pad($row->AppSite ?? '', $widths[0]);
            echo ' | ';
            echo str_pad($row->AppView ?? '', $widths[1]);
            echo ' | ';
            echo str_pad(number_format($row->NormalTimeMs, 2), $widths[2]);
            echo ' | ';
            echo str_pad(number_format($row->PreProcessTimeMs, 2), $widths[3]);
            echo ' | ';
            echo str_pad($row->ResultsMatch ?? '', $widths[4]);
            echo ' | ';
            echo str_pad($row->PerfDifference ?? '', $widths[5]);
            echo " |\n";
        }
        // Print bottom divider
        echo '|';
        for ($i = 0; $i < $colCount; $i++) {
            echo ' ' . str_repeat('-', $widths[$i]) . ' ';
            if ($i < $colCount - 1) echo '|';
        }
        echo "|\n";

        // Save HTML file
        try {
            $html = '<html><head><title>PHP Performance Summary Table</title><style>table{border-collapse:collapse;}th,td{border:1px solid #888;padding:4px;}th{background:#eee;}</style></head><body>';
            $html .= '<h2>PHP Performance Summary Table</h2>';
            $html .= '<table>';
            $html .= '<tr>';
            foreach ($headers as $h) {
                $html .= "<th>{$h}</th>";
            }
            $html .= '</tr>';
            foreach ($summaryRows as $row) {
                $html .= '<tr>';
                $html .= "<td>{$row->AppSite}</td>";
                $html .= "<td>{$row->AppView}</td>";
                $html .= "<td>" . number_format($row->NormalTimeMs, 2) . "</td>";
                $html .= "<td>" . number_format($row->PreProcessTimeMs, 2) . "</td>";
                $html .= "<td>{$row->ResultsMatch}</td>";
                $html .= "<td>{$row->PerfDifference}</td>";
                $html .= '</tr>';
            }
            $html .= '</table></body></html>';

            $outFile = $assemblerWebDirPath . DIRECTORY_SEPARATOR . 'php_perfsummary.html';
            file_put_contents($outFile, $html);
            echo "Performance summary HTML saved to: {$outFile}\n";
        } catch (\Exception $ex) {
            echo "Error saving performance summary HTML: {$ex->getMessage()}\n";
        }

        // Save JSON file
        try {
            $jsonFile = $assemblerWebDirPath . DIRECTORY_SEPARATOR . 'php_perfsummary.json';
            $jsonData = array_map(function($row) {
                return [
                    'AppSite' => $row->AppSite,
                    'AppFile' => $row->AppFile,
                    'AppView' => $row->AppView,
                    'Iterations' => $row->Iterations,
                    'NormalTimeMs' => round($row->NormalTimeMs, 2),
                    'PreProcessTimeMs' => round($row->PreProcessTimeMs, 2),
                    'OutputSize' => $row->OutputSize,
                    'ResultsMatch' => $row->ResultsMatch,
                    'PerfDifference' => $row->PerfDifference
                ];
            }, $summaryRows);
            $json = json_encode($jsonData, JSON_PRETTY_PRINT);
            file_put_contents($jsonFile, $json);
            echo "Performance summary JSON saved to: {$jsonFile}\n";
        } catch (\Exception $ex) {
            echo "Error saving performance summary JSON: {$ex->getMessage()}\n";
        }
    }
}

?>